File: Web/Engine/Utilities/Serialization/AxSerializationUtils.js
/**
 * @constructor
 */
function AxSerializationUtils()
{
}



// Static methods



/**
 * Serializes the given vector and writes the serialized data into the given stream
 * @param {AxStream} destinationStream Stream in which to write the serialized data
 * @param {AxVector2} vector The vector which is to be serialized
 */
AxSerializationUtils.SerializeVector2 = function(destinationStream, vector)
{
    destinationStream.WriteFloat32(vector.x);
    destinationStream.WriteFloat32(vector.y);
};

/**
 * Deserializes a vector from the given stream
 * @param {AxStream} sourceStream Stream with the serialized data
 * @return {AxVector2} The deserialized vector
 */
AxSerializationUtils.DeserializeVector2 = function(sourceStream)
{
    result = new AxVector2();
    result.x = sourceStream.ReadFloat32();
    result.y = sourceStream.ReadFloat32();
    return result;
};

/**
 * Serializes the given vector and writes the serialized data into the given stream
 * @param {AxStream} destinationStream Stream in which to write the serialized data
 * @param {AxVector3} vector The vector which is to be serialized
 */
AxSerializationUtils.SerializeVector3 = function(destinationStream, vector)
{
    destinationStream.WriteFloat32(vector.x);
    destinationStream.WriteFloat32(vector.y);
    destinationStream.WriteFloat32(vector.z);
};

/**
 * Deserializes a vector from the given stream
 * @param {AxStream} sourceStream Stream with the serialized data
 * @return {AxVector3} The deserialized vector
 */
AxSerializationUtils.DeserializeVector3 = function(sourceStream)
{
    result = new AxVector3();
    result.x = sourceStream.ReadFloat32();
    result.y = sourceStream.ReadFloat32();
    result.z = sourceStream.ReadFloat32();
    return result;
};

/**
 * Serializes the given vector and writes the serialized data into the given stream
 * @param {AxStream} destinationStream Stream in which to write the serialized data
 * @param {AxVector4} vector The vector which is to be serialized
 */
AxSerializationUtils.SerializeVector4 = function(destinationStream, vector)
{
    destinationStream.WriteFloat32(vector.x);
    destinationStream.WriteFloat32(vector.y);
    destinationStream.WriteFloat32(vector.z);
    destinationStream.WriteFloat32(vector.w);
};

/**
 * Deserializes a vector from the given stream
 * @param {AxStream} sourceStream Stream with the serialized data
 * @return {AxVector4} The deserialized vector
 */
AxSerializationUtils.DeserializeVector4 = function(sourceStream)
{
    result = new AxVector4();
    result.x = sourceStream.ReadFloat32();
    result.y = sourceStream.ReadFloat32();
    result.z = sourceStream.ReadFloat32();
    result.w = sourceStream.ReadFloat32();
    return result;
};

/**
 * Serializes the given matrix and writes the serialized data into the given stream
 * @param {AxStream} destinationStream Stream in which to write the serialized data
 * @param {AxMatrix} matrix The matrix which is to be serialized
 */
AxSerializationUtils.SerializeMatrix = function(destinationStream, matrix)
{
    destinationStream.WriteFloat32(matrix._11);
    destinationStream.WriteFloat32(matrix._12);
    destinationStream.WriteFloat32(matrix._13);
    destinationStream.WriteFloat32(matrix._14);
    destinationStream.WriteFloat32(matrix._21);
    destinationStream.WriteFloat32(matrix._22);
    destinationStream.WriteFloat32(matrix._23);
    destinationStream.WriteFloat32(matrix._24);
    destinationStream.WriteFloat32(matrix._31);
    destinationStream.WriteFloat32(matrix._32);
    destinationStream.WriteFloat32(matrix._33);
    destinationStream.WriteFloat32(matrix._34);
    destinationStream.WriteFloat32(matrix._41);
    destinationStream.WriteFloat32(matrix._42);
    destinationStream.WriteFloat32(matrix._43);
    destinationStream.WriteFloat32(matrix._44);
};

/**
 * Deserializes a matrix from the given stream
 * @param {AxStream} sourceStream Stream with the serialized data
 * @param {AxMatrix} result The deserialized matrix
 */
AxSerializationUtils.DeserializeMatrix = function(result, sourceStream)
{
    result._11 = sourceStream.ReadFloat32();
    result._12 = sourceStream.ReadFloat32();
    result._13 = sourceStream.ReadFloat32();
    result._14 = sourceStream.ReadFloat32();
    result._21 = sourceStream.ReadFloat32();
    result._22 = sourceStream.ReadFloat32();
    result._23 = sourceStream.ReadFloat32();
    result._24 = sourceStream.ReadFloat32();
    result._31 = sourceStream.ReadFloat32();
    result._32 = sourceStream.ReadFloat32();
    result._33 = sourceStream.ReadFloat32();
    result._34 = sourceStream.ReadFloat32();
    result._41 = sourceStream.ReadFloat32();
    result._42 = sourceStream.ReadFloat32();
    result._43 = sourceStream.ReadFloat32();
    result._44 = sourceStream.ReadFloat32();
};


/**
 * Serializes the given pixel format and writes the serialized data into the given stream
 * @param {AxStream} destinationStream Stream in which to write the serialized data
 * @param {AxPixelFormat} format The pixel format which is to be serialized
 */
AxSerializationUtils.SerializePixelFormat = function(destinationStream, format)
{
    destinationStream.WriteInt16(format.numChannels);
    destinationStream.WriteInt16(format.sizeInBytes);
    destinationStream.WriteData(format.channelId, format.numChannels);
    destinationStream.WriteData(format.channelBitCount, format.numChannels);
};

/**
 * Deserializes a pixel format from the given stream
 * @param {AxStream} sourceStream Stream with the serialized data
 * @return {AxPixelFormat} The deserialized pixel format
 */
AxSerializationUtils.DeserializePixelFormat = function(sourceStream)
{
    var result = new AxPixelFormat();

    result.numChannels = sourceStream.ReadInt16();
    result.sizeInBytes = sourceStream.ReadInt16();
    sourceStream.ReadData(result.channelId, result.numChannels);
    sourceStream.ReadData(result.channelBitCount, result.numChannels);

    return result;
};

File: Engine/Axis/Source/Engine/Utilities/Serialization/AxSerializationUtils.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxSerializationUtils.h"

#include "..\..\..\Tools\Streams\AxStream.h"

#include "..\..\Primitive\AxVector2.h"
#include "..\..\Primitive\AxVector3.h"
#include "..\..\Primitive\AxVector4.h"
#include "..\..\Primitive\AxMatrix.h"
#include "..\..\Primitive\AxBoundingVolume.h"
#include "..\..\Primitive\AxPixelFormat.h"


AxSerializationUtils::AxSerializationUtils(void)
{
}


AxSerializationUtils::~AxSerializationUtils(void)
{
}

void AxSerializationUtils::SerializeVector2(AxStream &destination, AxVector2 &vector)
{
	destination.WriteFloat32(vector.x);
	destination.WriteFloat32(vector.y);
}

AxVector2 AxSerializationUtils::DeserializeVector2(AxStream &source)
{
	AxVector2 result;
	result.x = source.ReadFloat32();
	result.y = source.ReadFloat32();
	return result;
}

void AxSerializationUtils::SerializeVector3(AxStream &destination, AxVector3 &vector)
{
	destination.WriteFloat32(vector.x);
	destination.WriteFloat32(vector.y);
	destination.WriteFloat32(vector.z);
}

AxVector3 AxSerializationUtils::DeserializeVector3(AxStream &source)
{
	AxVector3 result;
	result.x = source.ReadFloat32();
	result.y = source.ReadFloat32();
	result.z = source.ReadFloat32();
	return result;
}

void AxSerializationUtils::SerializeVector4(AxStream &destination, AxVector4 &vector)
{
	destination.WriteFloat32(vector.x);
	destination.WriteFloat32(vector.y);
	destination.WriteFloat32(vector.z);
	destination.WriteFloat32(vector.w);
}

AxVector4 AxSerializationUtils::DeserializeVector4(AxStream &source)
{
	AxVector4 result;
	result.x = source.ReadFloat32();
	result.y = source.ReadFloat32();
	result.z = source.ReadFloat32();
	result.w = source.ReadFloat32();
	return result;
}

void AxSerializationUtils::SerializeMatrix(AxStream &destination, AxMatrix &matrix)
{
	destination.WriteFloat32(matrix._11);
	destination.WriteFloat32(matrix._12);
	destination.WriteFloat32(matrix._13);
	destination.WriteFloat32(matrix._14);
	destination.WriteFloat32(matrix._21);
	destination.WriteFloat32(matrix._22);
	destination.WriteFloat32(matrix._23);
	destination.WriteFloat32(matrix._24);
	destination.WriteFloat32(matrix._31);
	destination.WriteFloat32(matrix._32);
	destination.WriteFloat32(matrix._33);
	destination.WriteFloat32(matrix._34);
	destination.WriteFloat32(matrix._41);
	destination.WriteFloat32(matrix._42);
	destination.WriteFloat32(matrix._43);
	destination.WriteFloat32(matrix._44);
}

void AxSerializationUtils::DeserializeMatrix(AxMatrix &result, AxStream &source)
{
	result._11 = source.ReadFloat32();
	result._12 = source.ReadFloat32();
	result._13 = source.ReadFloat32();
	result._14 = source.ReadFloat32();
	result._21 = source.ReadFloat32();
	result._22 = source.ReadFloat32();
	result._23 = source.ReadFloat32();
	result._24 = source.ReadFloat32();
	result._31 = source.ReadFloat32();
	result._32 = source.ReadFloat32();
	result._33 = source.ReadFloat32();
	result._34 = source.ReadFloat32();
	result._41 = source.ReadFloat32();
	result._42 = source.ReadFloat32();
	result._43 = source.ReadFloat32();
	result._44 = source.ReadFloat32();
}

void AxSerializationUtils::SerializePixelFormat(AxStream &destination, AxPixelFormat &format)
{
	destination.WriteInt16(format.numChannels);
	destination.WriteInt16(format.sizeInBytes);
	destination.WriteData(format.channelId, format.numChannels);
	destination.WriteData(format.channelBitCount, format.numChannels);
}

AxPixelFormat AxSerializationUtils::DeserializePixelFormat(AxStream &source)
{
	AxPixelFormat result;

	result.numChannels = source.ReadInt16();
	result.sizeInBytes = source.ReadInt16();
	source.ReadData(result.channelId, result.numChannels);
	source.ReadData(result.channelBitCount, result.numChannels);

	return result;
}

